### PR TITLE
opam: update to 2.3.0

### PR DESCRIPTION
**Pull request by kit-ty-kate**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y / Command Line Tools x.y.z

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by barracuda156** [comment]
It fails to build:
```
--->  Building opam
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_ocaml_opam/opam/work/opam-full-2.3.0" && /usr/bin/make -w all 
make: Entering directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_ocaml_opam/opam/work/opam-full-2.3.0'
dune build --profile=release --root .  --promote-install-files -- opam-installer.install opam.install
File "src/core/dune", line 13, characters 11-21:
13 |     (names opam_stubs)
                ^^^^^^^^^^
src/core/opamCommonStubs.c: In function 'opam_is_executable':
src/core/opamCommonStubs.c:55:9: error: implicit declaration of function 'faccessat'; did you mean 'access'? [-Wimplicit-function-declaration]
src/core/opamCommonStubs.c:55:19: error: 'AT_FDCWD' undeclared (first use in this function)
src/core/opamCommonStubs.c:55:19: note: each undeclared identifier is reported only once for each function it appears in
src/core/opamCommonStubs.c:55:38: error: 'AT_EACCESS' undeclared (first use in this function)
```

File: ocaml/opam/Portfile [hunk 1]
```diff
@@ -3,8 +3,9 @@
 PortSystem          1.0
 PortGroup           compiler_blacklist_versions 1.0
 PortGroup           github 1.0
+PortGroup           legacysupport
```
Comment (barracuda156) [inline ocaml/opam/Portfile:6]: It should be `legacysupport 1.1`.

And preferably also add:
```
# faccessat
legacysupport.newest_darwin_requires_legacy 13
```

**Comment by barracuda156** [comment]
If you prefer going with `legacysupport`, the build system should pass an ldflag for it. Otherwise linking fails:
```
--->  Building opam
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_ocaml_opam/opam/work/opam-full-2.3.0" && /usr/bin/make -w all 
make: Entering directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_ocaml_opam/opam/work/opam-full-2.3.0'
dune build --profile=release --root .  --promote-install-files -- opam-installer.install opam.install
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
File "src/tools/dune", line 28, characters 15-29:
28 |   (name        opam_installer)
                    ^^^^^^^^^^^^^^
Undefined symbols:
  "_faccessat", referenced from:
      _opam_is_executable in libopam_core_stubs.a(opam_stubs.o)
ld: symbol(s) not found
collect2: error: ld returned 1 exit status
File "_none_", line 1:
Error: Error while building custom runtime system
^Cmake: *** [build-opam] Error 130
```

File: ocaml/opam/Portfile [hunk 2]
```diff
@@ -62,9 +59,12 @@ if {[string match *gcc* ${configure.compiler}]} {
                     --without-mccs
 }
 
-if {${os.platform} eq "darwin" && ${os.major} < 11} {
-    # Matching ocaml 1.1 PG setting.
-    # See: https://github.com/ocaml/dune/issues/8941
-    build.env-append \
-                    DUNE_CONFIG__COPY_FILE=portable
+if {${os.platform} eq "darwin" && ${os.major} < 14} {
+    known_fail      yes
```
Comment (barracuda156) [inline ocaml/opam/Portfile:63]: Well, no, this is not something acceptable.

Let me find how Opam passes linker flags, and I will make a patch.

**Comment by kit-ty-kate** [comment]
I figured out how to stop macports from trying to build on such old systems. It is much easier this way.

**Review by barracuda156 — CHANGES_REQUESTED** [review]
Please remove `known_fail`, it is a fixable problem.

I will make a patch soon.

File: ocaml/opam/Portfile (hunk 2 above)
Comment (kit-ty-kate) [reply ocaml/opam/Portfile:63]: As you wish. I do not personally have more time to spend on this.

Comment (barracuda156) [reply ocaml/opam/Portfile:63]: I do not think there is a pressing need to update `opam` this very moment and break it for several systems. 2.2.0 has been working fine so far, AFAIK.

**Comment by barracuda156** [comment]
@kit-ty-kate Could you please borrow the correct fix from here? https://github.com/barracuda156/powerpc-ports/commit/58ec27fb68bf1f2815230f57548bddf3c8c03310
This builds now.

**Comment by kit-ty-kate** [comment]
> @kit-ty-kate Could you please borrow the correct fix from here? [barracuda156@58ec27f](https://github.com/barracuda156/powerpc-ports/commit/58ec27fb68bf1f2815230f57548bddf3c8c03310) This builds now.

done. Thanks!

**Review by barracuda156 — APPROVED** [review]

**Comment by barracuda156** [comment]
@kit-ty-kate Turns out, `opam` seems to ignore a setting to use vendored deps and uses `ocaml-dune` if it is available in the PATH. Which is why 2.3.0 was building fine for me (I have fixed dune for old systems earlier). However if dune is not available, opam fetches its source and bootstraps it – and that is still broken, as it was earlier: https://github.com/ocaml/opam/issues/5770

So to make it work with bundled dune, a hack is needed to pass the patch to opam.

Here is how to do it:
```
post-extract {
    xinstall -d ${worksrcpath}/src_ext/patches/dune-local
    copy ${filespath}/fix-bundled-dune.patch \
                    ${worksrcpath}/src_ext/patches/dune-local
}
```
And the patch itself (copied from dune portfile):
```
From 6120c556f8de5cc837832fe0d46b4fd4f88618ba Mon Sep 17 00:00:00 2001
From: barracuda156 <vital.had@gmail.com>
Date: Fri, 20 Oct 2023 04:54:18 +0800
Subject: [PATCH] copyfile_stubs.c: define COPYFILE_CLONE in case it is
 undefined

Credit to Ali Caglayan <alizter@gmail.com>
Picked from https://github.com/ocaml/dune/pull/8942/commits/89671ada56e24af7654a5b57b9089501342a2e62
---
 otherlibs/stdune/src/copyfile_stubs.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git otherlibs/stdune/src/copyfile_stubs.c otherlibs/stdune/src/copyfile_stubs.c
index 09c9a1b14..6824c572f 100644
--- a/otherlibs/stdune/src/copyfile_stubs.c
+++ b/otherlibs/stdune/src/copyfile_stubs.c
@@ -15,6 +15,10 @@
 #include <string.h>
 #include <sys/syslimits.h>
 
+#ifndef COPYFILE_CLONE
+#define COPYFILE_CLONE (1<<24)
+#endif
+
 CAMLprim value stdune_copyfile(value v_from, value v_to) {
   CAMLparam2(v_from, v_to);
   caml_unix_check_path(v_from, "copyfile");
```
Notice, there is still a problem that dune is used opportunistically.

**Pull request merged**